onnxruntime/core/optimizer/matmul_add_fusion.cc

```diff
@@ -65,13 +68,70 @@ Status MatMulAddFusion::ApplyImpl(Graph& graph, bool& modified, int graph_level,
       continue;
     }
 
-    if (2 != matmul_a_shape->dim_size() || 2 != matmul_b_shape->dim_size()) {
-      // Gemm only support Matrix
-      continue;
+    bool need_reshape = matmul_a_shape->dim_size() != 2;
+    const auto& dim_n = matmul_b_shape->dim(1);
+    std::vector<int64_t> shape_values;
+    int64_t m = 0, k = 0, n = 0;
+    if (need_reshape) {
+      // Skip Attention pattern, AttentionFusion will handle it. In such case, there are 4 MatMul-Add pairs,
+      // 3 of them are following LN, the other one produces output which is added by LN's output.
+      const Node* parent_node = graph.GetProducerNode(matmul_input_defs[0]->Name());
+      if (attn_ln_nodes.count(parent_node) > 0 || attn_add_nodes.count(&next_node) > 0) {
+        continue;
+      }
+      if (parent_node && parent_node->OpType() == "LayerNormalization") {
+        unsigned int add_count = 0;
+        unsigned int matmul_count = 0;
+        unsigned int shape_count = 0;
+        const Node* ln_add_node = nullptr;
+        for (auto it = parent_node->OutputNodesBegin(); it != parent_node->OutputNodesEnd(); ++it) {
+          std::string op_type = (*it).OpType();
+          if (op_type == "Add") {
+            ln_add_node = &(*it);
+            add_count++;
+          } else if (op_type == "MatMul") {
+            matmul_count++;
+          } else if (op_type == "Shape") {
+            shape_count++;
+          }
+        }
+        if (add_count == 1 && matmul_count == 3 && shape_count == parent_node->GetOutputEdgesCount() - 4) {
+          size_t index = ln_add_node->InputDefs()[0]->Name() == parent_node->OutputDefs()[0]->Name() ? 1 : 0;
+          const Node* attn_add_node = graph.GetProducerNode(ln_add_node->InputDefs()[index]->Name());
+          if (attn_add_node && attn_add_node->OpType() == "Add") {
+            attn_ln_nodes.insert(parent_node);
+            attn_add_nodes.insert(attn_add_node);
+            continue;
+          }
+        }
+      }
+
+      // Logically we can use Shape-Concat to produce shape input for Reshape, to keep it simple, we require
+      // both inputs have concrete shape for now, we can add dynamic shape support in future.
+      bool is_concrete_shape = true;
+      for (int i = 0; i < matmul_a_shape->dim_size(); ++i) {
+        const auto& dim = matmul_a_shape->dim(i);
+        if (!utils::HasDimValue(dim)) {
+          is_concrete_shape = false;
+          break;
+        }
+        shape_values.emplace_back(dim.dim_value());
+      }
+      if (!is_concrete_shape) {
+        continue;
+      }
+      const auto& dim_k = matmul_b_shape->dim(0);
+      if (!utils::HasDimValue(dim_k) || !utils::HasDimValue(dim_n)) {
+        continue;
+      }
+      k = dim_k.dim_value();
+      n = dim_n.dim_value();
+      m = std::accumulate(shape_values.begin(), shape_values.end() - 1, static_cast<int64_t>(1),
+                          std::multiplies<int64_t>());
```

Comment:
IIUC we don't need to prepend a 1 to the A input dims if it is 1D as '1' will be the default output from accumulate if there are not items to iterate. Would be good to capture that in a comment given it's a little opaque.